### ISSUE TITLE
0.18 Upgrade plan

### ISSUE THREAD

**Issue by mgold** [mgold/elm-random-pcg#13]
Because this library is a dependency for `elm-test` (hey @rtfeldman), which we want working with 0.18 on release, it's not too early to think about the upgrade plan. The major concerns are that bitshift ops are being renamed and their arg is being swapped, and backticks are being removed. The rename is actually a huge blessing: not only do I like the new names better, they allow the two arg orders to be distinguished.

So here's the cleverness that went down on the `update-0.18` branch: I made a shim module with the three newly named functions, which I imported in Pcg.elm `as Bitwise`. Then as I changed over each function I was able to check that this command still gave me the same output:

```
> import Random.Pcg as R
> R.step (R.list 10 (R.int 0 100)) (R.initialSeed 628138) |> fst
[43,10,36,56,83,99,69,10,48,93] : List Int
```

Indeed I did mess it up once, and by changing incrementally and checking I was able to catch the bug easily. So the upshot of this is that the code looks a lot more like Elm 0.18, while still being completely valid Elm 0.17.

Anyone who's so bold, please try the new branch and see if you can find any discrepancies between it and what's released -- the intention is that it behaves identically. Or just give me some code review and make sure I didn't make any silly mistakes (like changing order of ops or not changing to the new function names).

Still TODO: `rem` is used in a few places with backticks. The function is going to change in 0.18, but hasn't yet, so I just have to keep an eye on it. And of course `andThen` needs args swapped, but that's easy and would break compatibility if I did it now.

**Comment by rtfeldman** [comment]
Nice! Good thought getting in there early with this.

**Comment by mgold** [comment]
Latest revision requires a slight change thanks to `Tuple`:

``` elm
import Random.Pcg as R
import Tuple
R.step (R.list 10 (R.int 0 100)) (R.initialSeed 628138) |> Tuple.first
> [43,10,36,56,83,99,69,10,48,93] : List Int
```

Published 4.0.1.

**Comment by mgold** [comment]
Note to self: next time pick ranges that don't include zero, and are small enough that you'd see any numbers that should be absent. See f0a22c9434b541f8c2a3c2885057f19d28448599.

**Comment by mordrax** [comment]
@mgold I'd like to test the 0.18 branch out, is the easiest way to do this:
1. clone the update-0.18 branch
2. symlink from the branch to my app's elm-stuff dir
3. in elm-package.json, refer to the linked dir
(basically as described here: http://stackoverflow.com/questions/28110810/using-local-packages)

or is there any other way to reference local dependencies?

**Comment by eeue56** [comment]
@mgold 

hey, to unblock elm-test being released for 0.18, I published 4.0.0, 4.0.1, 4.0.2 against the new package server for you.

The changes needed to publish them invovled swapping out the `fst` calls for the `first`. Can you force-push 4.0.0, 4.0.1 to use the renamed Tuple functions so they work for 0.18?

**Comment by mgold** [comment]
@eeue56 I have merged `update-0.18` into master and force-retagged 4.0.{0,1,2} to the merge commit (which is a direct descendent of the old 4.0.2). So are we good?

@mordrax You should be able to use the package with 0.18 normally now.

Any other concerns or can I close the issue?

**Comment by rtfeldman** [comment]
Seems like this can be closed!
